File: src/filecompare_app.py
import tkinter as tk
from tkinter import filedialog, messagebox, ttk
from dataclasses import dataclass
from pathlib import Path


@dataclass
class OperationRow:
    op_type: str
    source: str
    target: str
    reason: str
    conflict: str


class FileCompareApp(tk.Tk):
    def __init__(self) -> None:
        super().__init__()
        self.title("FileCompare - Preview-first File Organizer")
        self.geometry("1200x700")

        self.selected_folder: Path | None = None
        self.status_var = tk.StringVar(value="Ready: 폴더를 선택하고 미리보기를 생성하세요.")

        self._build_toolbar()
        self._build_table()
        self._build_footer()

    def _build_toolbar(self) -> None:
        toolbar = ttk.Frame(self, padding=10)
        toolbar.pack(fill=tk.X)

        ttk.Button(toolbar, text="폴더 선택", command=self.select_folder).pack(side=tk.LEFT, padx=4)
        ttk.Button(toolbar, text="미리보기 생성", command=self.generate_preview).pack(side=tk.LEFT, padx=4)
        ttk.Button(toolbar, text="실행", command=self.apply_plan).pack(side=tk.LEFT, padx=4)

        ttk.Label(
            toolbar,
            text="기본 정책: Preview -> Confirm -> Apply",
            foreground="#444",
        ).pack(side=tk.RIGHT, padx=8)

    def _build_table(self) -> None:
        frame = ttk.Frame(self, padding=(10, 0, 10, 0))
        frame.pack(fill=tk.BOTH, expand=True)

        columns = ("type", "source", "target", "reason", "conflict")
        self.table = ttk.Treeview(frame, columns=columns, show="headings")
        self.table.heading("type", text="Type")
        self.table.heading("source", text="Source")
        self.table.heading("target", text="Target")
        self.table.heading("reason", text="Reason")
        self.table.heading("conflict", text="Conflict")

        self.table.column("type", width=110, anchor=tk.W)
        self.table.column("source", width=340, anchor=tk.W)
        self.table.column("target", width=340, anchor=tk.W)
        self.table.column("reason", width=270, anchor=tk.W)
        self.table.column("conflict", width=120, anchor=tk.W)

        y_scroll = ttk.Scrollbar(frame, orient=tk.VERTICAL, command=self.table.yview)
        self.table.configure(yscrollcommand=y_scroll.set)

        self.table.pack(side=tk.LEFT, fill=tk.BOTH, expand=True)
        y_scroll.pack(side=tk.RIGHT, fill=tk.Y)

    def _build_footer(self) -> None:
        footer = ttk.Frame(self, padding=10)
        footer.pack(fill=tk.X)
        ttk.Label(footer, textvariable=self.status_var).pack(side=tk.LEFT)

    def select_folder(self) -> None:
        selected = filedialog.askdirectory(title="분석할 폴더를 선택하세요")
        if not selected:
            return
        self.selected_folder = Path(selected)
        self.status_var.set(f"선택된 폴더: {self.selected_folder}")

    def generate_preview(self) -> None:
        self.table.delete(*self.table.get_children())

        if self.selected_folder is None:
            messagebox.showinfo("알림", "먼저 폴더를 선택하세요.")
            return

        rows = self._sample_rows(self.selected_folder)
        for row in rows:
            self.table.insert("", tk.END, values=(row.op_type, row.source, row.target, row.reason, row.conflict))

        self.status_var.set(f"미리보기 생성 완료: {len(rows)}건")

    def apply_plan(self) -> None:
        count = len(self.table.get_children())
        if count == 0:
            messagebox.showwarning("실행 불가", "실행할 미리보기 항목이 없습니다.")
            return

        messagebox.showinfo(
            "안전 모드",
            "현재 버전은 미리보기 중심 MVP입니다.\n실제 파일 변경은 아직 비활성화되어 있습니다.",
        )
        self.status_var.set("안전 모드: 현재는 미리보기만 제공됩니다.")

    @staticmethod
    def _sample_rows(base: Path) -> list[OperationRow]:
        downloads = str(base)
        return [
            OperationRow(
                "MOVE",
                f"{downloads}/receipt_2025_11.pdf",
                f"{downloads}/Documents/receipt_2025_11.pdf",
                "Rule: ext=pdf -> Documents",
                "NONE",
            ),
            OperationRow(
                "RENAME",
                f"{downloads}/IMG_0001.JPG",
                f"{downloads}/2026-02-20_IMG_0001.JPG",
                "Rule: camera files normalize",
                "NONE",
            ),
            OperationRow(
                "DELETE",
                f"{downloads}/copy_movie.mkv",
                "(Recycle Bin)",
                "Duplicate hash match group #14",
                "PROTECTED",
            ),
        ]


if __name__ == "__main__":
    app = FileCompareApp()
    app.mainloop()

File: src/main/java/com/filecompare/app/FileCompareApp.java
package com.filecompare.app;

import com.filecompare.model.OperationRow;
import javafx.application.Application;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.ToolBar;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.stage.Stage;

public class FileCompareApp extends Application {

    private final ObservableList<OperationRow> rows = FXCollections.observableArrayList();

    @Override
    public void start(Stage stage) {
        stage.setTitle("FileCompare - Preview-first File Organizer");

        var root = new BorderPane();
        root.setPadding(new Insets(10));

        var statusLabel = new Label("Ready: 폴더를 선택하고 미리보기를 생성하세요.");
        var toolbar = createToolbar(statusLabel);
        root.setTop(toolbar);

        var table = createPreviewTable();
        root.setCenter(table);

        var footer = new HBox(statusLabel);
        footer.setPadding(new Insets(8, 0, 0, 0));
        root.setBottom(footer);

        var scene = new Scene(root, 1200, 700);
        stage.setScene(scene);
        stage.show();
    }

    private ToolBar createToolbar(Label statusLabel) {
        var selectFolderButton = new Button("폴더 선택");
        var generatePreviewButton = new Button("미리보기 생성");
        var applyButton = new Button("실행");

        selectFolderButton.setOnAction(event -> statusLabel.setText("샘플: D:/Downloads 폴더가 선택되었습니다."));
        generatePreviewButton.setOnAction(event -> {
            rows.setAll(sampleRows());
            statusLabel.setText("미리보기 생성 완료: " + rows.size() + "건");
        });
        applyButton.setOnAction(event -> statusLabel.setText("안전 모드: 현재는 미리보기만 제공됩니다."));

        var spacer = new HBox();
        HBox.setHgrow(spacer, Priority.ALWAYS);

        return new ToolBar(selectFolderButton, generatePreviewButton, applyButton, spacer,
                new Label("기본 정책: Preview -> Confirm -> Apply"));
    }

    private TableView<OperationRow> createPreviewTable() {
        var table = new TableView<OperationRow>();

        var typeColumn = new TableColumn<OperationRow, String>("Type");
        typeColumn.setCellValueFactory(new PropertyValueFactory<>("type"));
        typeColumn.setPrefWidth(100);

        var sourceColumn = new TableColumn<OperationRow, String>("Source");
        sourceColumn.setCellValueFactory(new PropertyValueFactory<>("source"));
        sourceColumn.setPrefWidth(330);

        var targetColumn = new TableColumn<OperationRow, String>("Target");
        targetColumn.setCellValueFactory(new PropertyValueFactory<>("target"));
        targetColumn.setPrefWidth(330);

        var reasonColumn = new TableColumn<OperationRow, String>("Reason");
        reasonColumn.setCellValueFactory(new PropertyValueFactory<>("reason"));
        reasonColumn.setPrefWidth(250);

        var conflictColumn = new TableColumn<OperationRow, String>("Conflict");
        conflictColumn.setCellValueFactory(new PropertyValueFactory<>("conflictStatus"));
        conflictColumn.setPrefWidth(140);

        table.getColumns().addAll(typeColumn, sourceColumn, targetColumn, reasonColumn, conflictColumn);
        table.setItems(rows);
        table.setPlaceholder(new Label("미리보기 데이터가 없습니다."));

        return table;
    }

    private ObservableList<OperationRow> sampleRows() {
        return FXCollections.observableArrayList(
                new OperationRow(
                        "MOVE",
                        "D:/Downloads/receipt_2025_11.pdf",
                        "D:/Archive/Documents/receipt_2025_11.pdf",
                        "Rule: ext=pdf -> Documents",
                        "NONE"
                ),
                new OperationRow(
                        "RENAME",
                        "D:/Downloads/IMG_0001.JPG",
                        "D:/Downloads/2026-02-20_IMG_0001.JPG",
                        "Rule: camera files normalize",
                        "NONE"
                ),
                new OperationRow(
                        "DELETE",
                        "D:/Downloads/copy_movie.mkv",
                        "(Recycle Bin)",
                        "Duplicate hash match group #14",
                        "PROTECTED"
                )
        );
    }

    public static void main(String[] args) {
        launch(args);
    }
}
